### ISSUE TITLE
Support processing the commits starting from a specific tag

### ISSUE THREAD

**Issue by landerss1** [orhun/git-cliff#655]
### Is there an existing issue or pull request for this?

- [X] I have searched the existing issues and pull requests

### Feature description

Currently, there's support for processing commits starting from the current or latest tag, or commits without any tag, but not starting from a specific tag. This would make the transition from unconventional commits to conventional commits  much smoother. Then, you could keep the old changelog as, e.g. `HISTORY.md`, and then start a new `CHANGELOG.md` based on commits after a specific tag.

### Desired solution

Introduce a new command-line argument that can be used to specify the tag that should be used to start fetching commits from.

### Alternatives considered

Can't think of any.

**Comment by orhun** [comment]
Hello!

Have you tried the git `range` option?

e.g.

```sh
git cliff v0.1.0..HEAD
```

Or even more shorter:

```sh
git cliff v2.2.1..
```

Other examples:

```sh
git cliff 4c7b043..a440c6e
git cliff 4c7b043..HEAD
git cliff HEAD~2..
```

I think this is what you are looking for :)

**Comment by landerss1** [comment]
Ah, Thanks. I saw the range CLI argument in the docs, but there was no description or examples. I have tried it now, and it solves my problem, but you have to go one tag further back than expected, so you have to specify `range start=<desired-tag>-1`. Is this what's to be expected?

**Comment by orhun** [comment]
I think yes, since [git ranges](https://git-scm.com/docs/git-range-diff) work like that. It is handled by the underlying git library (`git2`).

I expanded the examples in d4512521fbcfb971c94aa7794d78bced1ddec7a1

**Comment by lermontex** [comment]
@orhun, as far as I understand, git ranges will conflict with the --unreleased, --latest, etc. flags, so git ranges is not entirely suitable, since it will not allow flexible configuration

It would be great if it was possible to specify the tag from which to start the story

There is an initial_tag parameter in the configuration, maybe it’s worth using it?

https://git-cliff.org/docs/configuration/bump/#initial_tag

**Comment by orhun** [comment]
> as far as I understand, git ranges will conflict with the --unreleased, --latest, etc. flags, so git ranges is not entirely suitable, since it will not allow flexible configuration

Yeah, using those flags will override the commit range.

Can you give me an example case (i.e. git history) to make this matter more clear? I'm happy to implement it as long as we have the correct behavior with [test fixtures](https://github.com/orhun/git-cliff/tree/main/.github/fixtures#fixtures) in place.

> There is an initial_tag parameter in the configuration, maybe it’s worth using it?

`initial_tag` option is related to the bump configuration and I don't think it is suitable for this case.

**Comment by lermontex** [comment]
@orhun, sorry for the delay in replying

A simple example: we have a repository that does not adhere to Conventional Commits (or partially adheres). We want to display history only from a certain tag (when we decided that we will strictly follow the rules)

At the moment, it can only be done this way:
`git-cliff v1.0.0.. --bump --output "CHANGELOG.md"`

It would be handy if the initial tag `v1.0.0` could be fixed in the configuration, rather than having to be specified each time `git-cliff` is run

**Comment by orhun** [comment]
Ah I see, you basically want a new configuration option for the commit range.

It seems like an overly specific case though, maybe instead we can support an environment variable such as `GIT_CLIFF__RANGE` . Would that be useful for your case?

**Comment by lermontex** [comment]
@orhun, unfortunately, I don't understand how I can use an environment variable in this case and how it will be better than the current option, when we specify the tag directly in the console command

Basically, I want to configure git cliff so that the history always starts with a certain GIT tag, so I would like to worry about this only once

If this seems like a rare use case, here is another example: we want to have several branches (1.0, 1.1, ... 2.0) and each of them should have its own changelog:

```
1.0 - CHANGELOG-1.0.md
1.1 - CHANGELOG-1.1.md
...
2.0 - CHANGELOG-2.0.md
```

Something like here:
https://github.com/symfony/symfony

If it were possible to specify the starting tag, we could configure each branch using cliff.toml so that it does not include changes from previous branches

Hopefully this is a more common use case :)

**Comment by orhun** [comment]
I see... then we can probably add a new option to the configuration file as follows then:

```toml
[git]
range = "1.0.0" # new option
output = "CHANGELOG-1.0.md" # this is already supported
```

How does this look to you?

**Comment by Cyclonit** [comment]
What about two options `git.gnore_before` and `git.gnore_after` (or `include_after` and `include_before`) instead? That would give you all the flexibility you might need.

**Comment by lermontex** [comment]
@orhun, yeah that would be great! Basically we could just allow the range parameter to be set directly in the config file, then we could use something like this:

```
[git]
range = "v1.0.0..v1.1.0"
output = "CHANGELOG-1.0.md" # this is already supported
```

@Cyclonit, note that in the range parameter you can specify a range of values, so most likely there is no need to use 2 config parameters

```
git cliff 4c7b043..a440c6e
git cliff 4c7b043..HEAD
git cliff HEAD~2..
git cliff v2.2.1..
git cliff v0.1.0..HEAD
```

https://git-cliff.org/docs/usage/examples/

Although, there is currently no way to specify whether the specified range should include the specified commits/tags? Unfortunately I couldn't find an example

**Comment by orhun** [comment]
> Basically we could just allow the range parameter to be set directly in the config file, then we could use something like this:

Great! I might implement that in the future - or I'm happy to review the PR if somebody wants to give it a shot!

> there is currently no way to specify whether the specified range should include the specified commits/tags? Unfortunately I couldn't find an example

I'm not sure if I got your question, can you elaborate a bit?

**Comment by lermontex** [comment]
@orhun, I mean that using the `range` parameter the boundary commits will be included in the history (as far as I understand):

`git cliff 4c7b043..a440c6e`

There is currently no way to exclude commits `4c7b043` and `a440c6e` from the history

Personally, this is not a issue for me, it's just that the options @Cyclonit suggested (`ignore_before`/`ignore_after` (or `include_after`/`include_before`) assume that the boundary commits will be excluded from the history. Just wanted to point this out

**Comment by jankatins** [comment]
Just to add to the voices asking for it in the config file: we run git cliff in a pipeline which is supplied by our infra team. Adding a env var or adding a cli command argument is hard to do, so a item in the config file would be highly appreciated :-)

**Comment by orhun** [comment]
I would appreciate a PR on this!

**Comment by cds-amal** [comment]
## Food for thought, apologies upfront for the indigestion

_I've managed a popular Open Source tool in the past, and some of that pain influenced this bit of research. I'm new to git-cliff and found the range-selection flags confusing the first time I tried to reason about what would actually end up in my changelog. So I did the thing you do in that situation: I worked out, flag by flag, what each one does to the underlying set of commits. Once that picture came into focus, a proposal for this issue fell out naturally, and I think it also explains why the single-option proposals so far (`ignore_before`, `range` in config, `start_after`, etc.) kept running into the same arguments in the thread._

## Setting the scene: what does a git-cliff run actually compute?

Every git-cliff run, stripped of its grouping and templating, does one thing: it picks a contiguous interval of commits from the repo's history and hands it off to the rest of the pipeline. The flags that exist today (`--latest`, `--unreleased`, `--current`, `--bump`, the positional `<A>..<B>`) are all shortcuts that fill in the two endpoints of that interval.

That is the part that took me a while to see clearly, so let me be specific: **every flag is a named way of picking `(left endpoint, right endpoint)`**, with an inclusivity choice silently baked in for each side.

Here is the decomposition, using standard math interval notation (`[x, y]` means both `x` and `y` are included; `(x, y]` means `x` is excluded and `y` is included; and so on). Names like `last_tag` refer to tags resolved from the repo at invocation time.

| | Flag / invocation | Left endpoint         | Right endpoint                                      | Interval               |
| - | ----------------- | --------------------- | --------------------------------------------------- | ---------------------- |
|1| (no flags)        | first commit          | HEAD                                                | `[first, HEAD]`        |
|2| `--unreleased`    | last tag              | HEAD                                                | `(last_tag, HEAD]`     |
|3| `--latest`        | second-to-last tag    | last tag                                            | `(prev_tag, last_tag]` |
|4| `--current`       | tag before current    | currently checked-out tag                           | `(prev_tag, current]`  |
|5| `--bump`          | last tag              | HEAD (with a synthesized next-version tag applied)  | `(last_tag, HEAD]`     |
|6| `<A>..<B>`        | A                     | B                                                   | `(A, B]`               |

> **N.B.** Every tag-based row is left-exclusive. That is not an accident; git's `A..B` syntax *is* left-exclusive, and git-cliff inherits that convention from the underlying `git2` library. The `(no flags)` row is the one exception, and it is closed on the left only because the root commit has no ancestor to exclude. So if you want to include a specific tag's commit in your output using today's flags, you have to know the trick of going one tag further back than you'd expect; the maintainer's first comment on this issue is exactly that observation.

## The feature request, stated plainly

The ask is for a way to say "start the changelog at tag X" that can live in config so it gets set once and forgotten (multi-branch changelogs and CI pipelines that cannot vary their CLI invocation are the concrete motivators in the thread).

The natural first move, which the thread explored, is to add a single config option. Call it `ignore_before`, or `range = "X.."`, or `start_after`; I tried each of these while writing this up and watched each one fail in the same way. The failure, I think, has a single cause.

Every name chosen baked an inclusivity decision into the name, and each decision was wrong for some real user:

- `ignore_before = "v1.0.0"` reads in English as inclusive (`[v1.0.0, …]`); implemented that way, v1.0.0 renders sparsely because its body commits fall below the floor. Flip to exclusive `(v1.0.0, …]` and the name no longer means what it says.
- `range = "v1.0.0.."` sidesteps the English question but inherits git's left-exclusive convention, dropping us back into the "go one tag further back" confusion that opened the issue.

None of these arguments was wrong. They are all symptoms of the same thing: **trying to compress two orthogonal design choices (left endpoint, right endpoint) plus two other orthogonal inclusivity choices into a single-option name**. The arguments will keep surfacing as long as the surface keeps hiding those choices.

## Proposal: put the four choices into four options

If the confusion comes from hiding inclusivity behind names, the fix, IMHO, is to put it back on the surface. Four options, one per (side × inclusivity) combination, with the choice in the option name itself:

| | CLI flag          | Config key          | Meaning                      | Contribution  |
| - | ----------------- | ------------------- | ---------------------------- | ------------- |
|1| `--start-at X`    | `start_at = "X"`    | include `X`; walk forward    | `[X, ·`       |
|2| `--start-after X` | `start_after = "X"` | exclude `X`; walk forward    | `(X, ·`       |
|3| `--end-at Y`      | `end_at = "Y"`      | include `Y`; walk back       | `·, Y]`       |
|4| `--end-before Y`  | `end_before = "Y"`  | exclude `Y`; walk back       | `·, Y)`       |

The pattern: `*_at` is always inclusive, and its positional-word sibling (`*_after` for the left side, `*_before` for the right side) is always exclusive. Within each pair, exactly one may be set. The two pairs are independent, so all four math intervals are expressible: `[A, B]`, `(A, B]`, `[A, B)`, `(A, B)`. Unspecified sides fall back to the existing defaults (left = first commit, inclusive; right = HEAD, inclusive).

> **Remark.** The naming discipline is doing real work here. "At" is naturally inclusive in English ("on March 5th at noon" includes noon), while "after" and "before" are naturally exclusive ("the day after Christmas" is December 26, not the 25th). We get the inclusivity story for free once we trust that convention.

## What the existing flags become

Nothing about the existing flags changes. They keep doing exactly what they do today. But the decomposition above lets us document what each one computes in the vocabulary of the new options, which doubles as a rosetta stone for anyone migrating to the explicit form:

| | Legacy         | Equivalent long form                                          |
| - | -------------- | ------------------------------------------------------------- |
|1| (no flags)     | `--end-at HEAD` (left defaults to first commit)               |
|2| `--unreleased` | `--start-after <last_tag> --end-at HEAD`                      |
|3| `--latest`     | `--start-after <prev_tag> --end-at <last_tag>`                |
|4| `--current`    | `--start-after <prev_tag> --end-at <current_tag>`             |
|5| `--bump`       | `--start-after <last_tag> --end-at HEAD` (plus version synthesis) |
|6| `<A>..<B>`     | `--start-after A --end-at B`                                  |

> **N.B.** Mixing a legacy shortcut flag with any of the new endpoint options is a user error and should be rejected with a clear message. The two styles do not compose cleanly (what would `--latest --end-before v1.0.0` even mean?); forcing the user to pick one avoids a class of bug reports.

## Worked examples

I'll use this fixture throughout: `A - B(v0.5.0) - C - D(v1.0.0) - E - F(v1.1.0) - G - H(HEAD)`. Each tag sits on the lettered commit shown, and HEAD is on H.

| | Invocation                                                    | Interval  | Sections rendered                                          |
| - | ------------------------------------------------------------- | --------- | ---------------------------------------------------------- |
|1| `git cliff`                                                   | `[A, H]`  | v0.5.0: A, B; v1.0.0: C, D; v1.1.0: E, F; unreleased: G, H |
|2| `git cliff --latest`                                          | `(D, F]`  | v1.1.0: E, F                                               |
|3| `git cliff --start-after v0.5.0` *(the canonical issue case)* | `(B, H]`  | v1.0.0: C, D; v1.1.0: E, F; unreleased: G, H               |
|4| `git cliff --start-at v1.0.0`                                 | `[D, H]`  | v1.0.0: D; v1.1.0: E, F; unreleased: G, H                  |
|5| `git cliff --start-after v0.5.0 --end-before v1.1.0`          | `(B, F)`  | v1.0.0: C, D; unreleased: E                                |
|6| `git cliff --end-at v1.0.0`                                   | `[A, D]`  | v0.5.0: A, B; v1.0.0: C, D                                 |

Row 3 is @landerss1's original ask: "start after v0.5.0, the last non-conventional release." No knowledge of git's left-exclusive convention required.

Contrast row 4: `--start-at v1.0.0` makes v1.0.0 the floor, so C (one of v1.0.0's body commits) falls below and the section renders with just D. Sometimes what users want, more often not; the point is that the name makes it visible rather than surprising someone later.

## Precedence and conflicts

- Within a pair, at most one option may be set. `start_at` and `start_after` together is a configuration error; same for `end_at` and `end_before`.
- A legacy shortcut flag may not be combined with any of the new endpoint options.
- CLI endpoint options override config endpoint options on the same side (so a shared team config can be locally overridden without editing the file).
- Any endpoint whose revision cannot be resolved is a hard error at config load / arg parse time, not a deferred one.
- If the evaluated interval is empty (`--start-at X --end-before X` is the simplest example; a floor above the computed right endpoint is another), warn on stderr and render an empty body. The header and footer still apply if the template defines them.

## Proposed implementation

The change lives in one function. `determine_commit_range` at `git-cliff/src/lib.rs:49` is a pure string-producer that returns `Option<String>`; everything downstream (release grouping, processing, templating, remote enrichment) just consumes the resulting commit list. The four options compose into a git range inside that function, using the `^` suffix trick for inclusive endpoints:

| | Left option       | Right option      | Emitted git range              |
| - | ----------------- | ----------------- | ------------------------------ |
|1| `--start-after A` | `--end-at B`      | `A..B` (git's default)         |
|2| `--start-at A`    | `--end-at B`      | `A^..B`                        |
|3| `--start-after A` | `--end-before B`  | `A..B^`                        |
|4| `--start-at A`    | `--end-before B`  | `A^..B^`                       |
|5| (default left)    | `--end-at B`      | `B` (all ancestors of B)       |
|6| `--start-after A` | (default right)   | `A..HEAD`                      |
|7| `--start-at A`    | (default right)   | `A^..HEAD`                     |
|8| (default left)    | (default right)   | `None` (walk everything)       |

Edge cases:

- If `A` is the root commit, `A^` does not resolve. Fall back to "no left bound", which is what `--start-at <root>` naturally means anyway.
- If `B` is unreachable from HEAD (lives on a side branch never merged in), the emitted range parses but comes back empty from `git2`. Warn rather than error.

Only the range-picker moves; processing, release assembly, templating, and the remote integrations need no changes.

## Companion proposal: `--dry-run`

A flag that prints the computed interval and exits. Lets users verify the selection before rendering, makes the scenarios above runnable as checks, and lets CI pipelines ask "is there anything here to release?" without a full render.

```
$ git cliff --dry-run
[A, H]
8 commits

$ git cliff --start-after v0.5.0 --dry-run
(B, H]
6 commits

$ git cliff --latest --dry-run
(D, F]
2 commits
```

Orthogonal to the endpoint API; the name is negotiable (`--no-exec` is already taken for a narrower purpose).

**Comment by cds-amal** [comment]
The endpoint surface I mentioned above landed in #1489. If approved, with `CommitRange` modeled as two optional `Endpoint`s flowing through a `transform -> resolve -> execute` pipeline, the **range layer** has become a composable seam. A few directions that fall out naturally; each is a pure transform on `CommitRange`, no fan-out across CLI / config / walker:

| | Operator | What it does | When you'd reach for it |
|--|---|---|---|
| 1 | `nth_tag(N)`, `since(<date>)`, `until(<date>)`, `branch(<name>)` | New endpoint kinds; each resolves to an `Endpoint` the same way `--start-at <tag>` does today | Time-based windows, "the last 3 releases", branch-tip ranges |
| 2 | `exclude(<commits-or-predicate>)` | Drop specific commits (or anything matching a predicate) before rendering | Dependabot churn, revert pairs, fixup commits, known-noisy authors :sweat_smile: |
| 3 | `union(r1, r2)`, `intersect(r1, r2)`, `diff(r1, r2)` | Set ops over two ranges | "everything on `main` since v2.0.0, union everything on `release/v3.x`" |
| 4 | `partition(by = author / scope / month)` | Split a range into a structured set, feeding a per-group changelog template | Per-author release notes, monthly digests, scope-based subsections |

> **N.B.** Rows 1–3 live in the **range layer** (operations on `CommitRange`). Row 4 bridges into the **commit-processing layer** that already exists (parsers, link rules, grouping). The seam between the two is the `CommitRange` value the PR introduces.

Read top-to-bottom, a pipeline becomes an intent statement:

```
range(--start-after v1.0.0)
  | exclude(authors = ["dependabot[bot]"])
  | filter(scope = "ui")
  | group_by(commit.month)
```

The first two lines are range-layer transforms; the last two live in the commit-processing layer that exists today. The seam stays narrow on purpose: each new operator still needs a surface (a flag, a config key, possibly both), but the architectural cost is one transform on `CommitRange` rather than a fresh path through CLI parsing, repo traversal, and release assembly. "Last 3 releases, dropping reverts, grouped by scope" stops being a feature with three new code paths and starts being a small composition over existing ones.
